### PR TITLE
Initial implementation of firebase-encoders-proto

### PR DESCRIPTION
**Pull request by vkryachko**

### CODE REVIEW

File: encoders/firebase-encoders-proto/src/main/java/com/google/firebase/encoders/proto/LengthCountingOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import androidx.annotation.NonNull;
+import java.io.OutputStream;
+
+/** OutputStream that only keeps track of the number of bytes written into it. */
+final class LengthCountingOutputStream extends OutputStream {
+  private long length = 0;
+
+  @Override
+  public void write(int b) {
+    length++;
```
[the remark this answers is not in the document]
Comment (vkryachko) [reply encoders/firebase-encoders-proto/src/main/java/com/google/firebase/encoders/proto/LengthCountingOutputStream.java:26]: That's what I thought too, but this method actually represents one byte with a 4 byte int, here's its javadoc:

```java
  /**
     * Writes the specified byte to this output stream. The general
     * contract for <code>write</code> is that one byte is written
     * to the output stream. The byte to be written is the eight
     * low-order bits of the argument <code>b</code>. The 24
     * high-order bits of <code>b</code> are ignored.
```

so +1 is actually correct

File: encoders/firebase-encoders-proto/src/main/java/com/google/firebase/encoders/proto/LengthCountingOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    length += b.length;
+  }
+
+  @Override
+  public void write(@NonNull byte[] b, int off, int len) {
+    if ((off < 0)
+        || (off > b.length)
+        || (len < 0)
+        || ((off + len) > b.length)
+        || ((off + len) < 0)) {
```
[the remark this answers is not in the document]
Comment (vkryachko) [reply encoders/firebase-encoders-proto/src/main/java/com/google/firebase/encoders/proto/LengthCountingOutputStream.java:40]: That's a great observation, I took this check directly from the jdk implementation, I wonder why it is this way. I can change it if you like and I don't feel strongly either way.

[the remark this answers is not in the document]
Comment (rlazo) [reply encoders/firebase-encoders-proto/src/main/java/com/google/firebase/encoders/proto/LengthCountingOutputStream.java:40]: That's OK, let's go the way it is.

**Review by rlazo — APPROVED** [review]

**Pull request merged**